test/gateway/index.js

```diff
@@ -0,0 +1,65 @@
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+const API = require('../../src/http')
+
+describe('HTTP Gateway', () => {
+  let http = {}
+  let gateway
+
+  before((done) => {
+    http.api = new API()
+
+    http.api.start(true, () => {
+      gateway = http.api.server.select('Gateway')
+      done()
+    })
+  })
+
+  after((done) => {
+    http.api.stop((err) => {
+      expect(err).to.not.exist()
+      done()
+    })
+  })
+
+  describe('/ipfs/* route', () => {
+    it('returns 400 for request without argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('400 for request with invalid argument', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/invalid'
+      }, (res) => {
+        expect(res.statusCode).to.equal(400)
+        expect(res.result.Message).to.be.a('string')
+        done()
+      })
+    })
+
+    it('valid hash', (done) => {
+      gateway.inject({
+        method: 'GET',
+        url: '/ipfs/QmT78zSuBmuS4z925WZfrqQ1qHaJ56DQaTfyMUF7F8ff5o'
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.rawPayload).to.deep.equal(Buffer.from('hello world' + '\n'))
+        expect(res.payload).to.equal('hello world' + '\n')
+        done()
+      })
+    })
+  })
```

Comment:
I believe that is missing to merge this PR is more tests.

There should be tests for:
- [ ] Fetching a file with a mime-type that is not text (i.e a pdf or an image)
- [ ] Loading a webpage (one for each type of index.html)
- [ ] Loading a webpage from a pagh (QmHASH/a/b/c -> loads the webpage in that path)

And other tests for future PR (I don't want to block this any longer)

- [ ] CORS tests (make sure config is respected)

Reply:
@ya7ya what do you think?

Reply:
@diasdavid I'm working on tests now, It took me a minute to figure out aegir :D and interface-ipfs-core. and i'm trying to follow that to complete the tests for the gateway.

I'll add the tests you mentioned above and check the Error with libp2p.dial you mentioned aswell. I'll try to get this done As soon as possible 👍